core/src/main/scala/org/apache/spark/sql/rapids/tool/ClusterTagPropHandler.scala

```diff
@@ -0,0 +1,110 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.tool
+
+import com.nvidia.spark.rapids.tool.planparser.DatabricksParseHelper
+
+import org.apache.spark.scheduler.SparkListenerJobStart
+import org.apache.spark.sql.rapids.tool.util.CacheablePropsHandler
+
+// Maintains and handles cluster tags for a Spark application.
+trait ClusterTagPropHandler extends CacheablePropsHandler {
+  // clusterTags can be redacted so try to get ClusterId and ClusterName separately
+  var clusterTags: String = ""
+  var clusterTagClusterId: String = ""
+  var clusterTagClusterName: String = ""
+
+  // A flag to indicate whether the eventlog being processed is an eventlog from Photon.
+  var isPhoton = false
+
+  // Flag used to indicate that the App was a Databricks App.
+  def isDatabricks: Boolean = {
+    clusterTags.nonEmpty && clusterTagClusterId.nonEmpty && clusterTagClusterName.nonEmpty
+  }
+
+  private def updateClusterTagsFromSparkProperties(): Unit = {
+    // Update the clusterTags if any.
+    if (clusterTags.isEmpty) {
+      clusterTags = sparkProperties.getOrElse(DatabricksParseHelper.PROP_ALL_TAGS_KEY, "")
+    }
+    if (clusterTagClusterId.isEmpty) {
+      clusterTagClusterId =
+        sparkProperties.getOrElse(DatabricksParseHelper.PROP_TAG_CLUSTER_ID_KEY, "")
+    }
+    if (clusterTagClusterName.isEmpty) {
+      clusterTagClusterName =
+        sparkProperties.getOrElse(DatabricksParseHelper.PROP_TAG_CLUSTER_NAME_KEY, "")
+    }
+    // update the photon Flag
+    if (isDatabricks) {
+      // update photonFlag
+      isPhoton ||= DatabricksParseHelper.isPhotonApp(sparkProperties)
+      // set the spark version
+      sparkVersion = DatabricksParseHelper.getSparkVersion(sparkProperties)
+    }
+  }
+
+  override def handleJobStartForCachedProps(event: SparkListenerJobStart): Unit = {
+    super.handleJobStartForCachedProps(event)
+    // If the confs are set after SparkSession initialization, it is captured in this event.
+    val dbFlagPreprocess = isDatabricks
+    if (!isDatabricks) {
+      // Add the clusterTags to SparkProperties if any
+      Seq(DatabricksParseHelper.PROP_ALL_TAGS_KEY,
+          DatabricksParseHelper.PROP_TAG_CLUSTER_ID_KEY,
+          DatabricksParseHelper.PROP_TAG_CLUSTER_NAME_KEY
+      ).foreach { tagKey =>
+        sparkProperties += (tagKey -> event.properties.getProperty(tagKey, ""))
+      }
+    }
+
+    if (!dbFlagPreprocess && isDatabricks) {
```

Comment:
In which step might `isDatabricks` get updated so that we need `dbFlagPreprocess`?

Reply:
I had the same question.

Reply:
My understanding of this code is: If either one or more clusterTags config is set after the SparkSession initialization then the flow would be:

`dbFlagPreprocess` = false ( assuming some/allclusterTags are not set)

`if (!isDatabricks)` //  set the clusterTag properties if provided, this would update sparkProperties. So if all clustertags are set, then the next time isDatabricks should return true.

` if (!dbFlagPreprocess && isDatabricks)` ->  This is to set isPhoton and sparkVersion